alea/ces_source.py

```diff
@@ -0,0 +1,234 @@
+        dtype = [
+            ("ces", float),
+            ("source", int),
+        ]
+        ret = np.zeros(n_events, dtype=dtype)
+        ret["ces"] = self._pdf_histogram.get_random(n_events)
+        return ret
+
+    def compute_pdf(self):
+        self.build_histogram()
+        Source.compute_pdf(self)
+
+    def pdf(self, *args):
+        # override the default interpolation method in blueice (RegularGridInterpolator)
+        if not self.pdf_has_been_computed:
+            raise PDFNotComputedException(
+                "%s: Attempt to call a PDF that has not been computed" % self
+            )
+
+        method = self.config["pdf_interpolation_method"]
+
+        if method == "linear":
+            if not hasattr(self, "_pdf_interpolator"):
+                # First call:
+                # Construct a linear interpolator between the histogram bins
+                self._pdf_interpolator = interp1d(
+                    self._pdf_histogram.bin_centers,
+                    self._pdf_histogram.histogram,
+                )
+            bcs = self._pdf_histogram.bin_centers
+            clipped_data = np.clip(args, bcs.min(), bcs.max())
+
+            return self._pdf_interpolator(np.transpose(clipped_data))
+
+        elif method == "piecewise":
+            return self._pdf_histogram.lookup(*args)
+
+        else:
+            raise NotImplementedError("PDF Interpolation method %s not implemented" % method)
+
+    def set_dtype(self):
+        self.dtype = [
+            ("ces", float),
+            ("source", int),
+        ]
+
+
+class CESMonoenergySource(CESTemplateSource):
+    def _load_inputs(self):
+        self.ces_space = self.config["analysis_space"][0][1]
+        self.max_e = np.max(self.ces_space)
+        self.min_e = np.min(self.ces_space)
+        self.mu = self.config["peak_energy"]
+
+    def _load_true_histogram(self):
+        number_of_bins = int((self.max_e - self.min_e) / MINIMAL_ENERGY_RESOLUTION)
+        h = Hist1d(
+            data=np.repeat(self.mu, 1),
+            bins=number_of_bins,
+            range=(self.min_e, self.max_e),
+        )
+        h.histogram = h.histogram.astype(np.float64)
+        self.config["smearing_model"] = "mono_" + self.config["smearing_model"]
+        return h
+
+
+class CESFlatSource(CESTemplateSource):
+    def _load_inputs(self):
+        self.ces_space = self.config["analysis_space"][0][1]
+        self.max_e = np.max(self.ces_space)
+        self.min_e = np.min(self.ces_space)
+
+    def _load_true_histogram(self):
+        number_of_bins = int((self.max_e - self.min_e) / MINIMAL_ENERGY_RESOLUTION)
+        h = Hist1d(
+            data=np.linspace(self.min_e, self.max_e, number_of_bins),
+            bins=number_of_bins,
+            range=(self.min_e, self.max_e),
+        )
+        h.histogram = h.histogram.astype(np.float64)
+        return h
+
+    def _transform_histogram(self, h: Hist1d):
+        # Only efficiency is applicable for flat source
```

Comment:
Disagree-- you could have an efficiency defined in true energy and then a smearing on top, I think?

Reply:
IMO the efficiency should be the probability to detect an event given a certain initial recoil energy. This is what we use for WIMPs, and it is a well-defined probability (that does include the effects you list here as important inputs!). 
I think attempting to apply efficiencies based on reconstructed variables on the basis that those reconstructed variables approximate (but does not give) the underlying number of quanta that are important for the probability to be detected

Reply:
Hi Knut, to clarify, the efficiency here is only projected in CES space. For S2 threshold and S1 efficiency, we get the CES projected values during the band fitting. For cut acceptance which is data-driven, the values are naturally only in CES. Defining everything in true energy is possible but the gain is limited. It will make every efficiency study entangled with the band fitting which is not necessary for lower or other studies performed in CES.